### PR TITLE
enum case

### PR DESCRIPTION
**Pull request by a-zakir**

### CODE REVIEW

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h [hunk 1]
```diff
@@ -22,21 +22,20 @@
 
 namespace Antares::Solver::Visitors
 {
-class TimeIndex
+enum class TIME_STRUCTURE_TYPE : int
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h]:
```suggestion
enum class TimeIndex : unsigned int
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h [hunk 2]
```diff
@@ -22,21 +22,20 @@
...
-    bool IsScenarioVarying() const;
-    TimeIndex Connect(const TimeIndex& other) const;
-
-private:
-    bool time_varying_ = false;
-    bool scenario_varying_ = false;
+    CONSTANT = 0,
+    VARYING_IN_TIME_ONLY = 1,
+    VARYING_IN_SCENARIO_ONLY = 2,
+    BOTH = 3
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h]:
```suggestion
    VARYING_IN_TIME_ONLY = 1,
    VARYING_IN_SCENARIO_ONLY = 2,
    VARYING_IN_TIME_AND_SCENARIO = 3
```

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 3]
```diff
@@ -363,40 +363,52 @@ BOOST_FIXTURE_TEST_CASE(simple_constant_expression, Registry<Node>)
...
                     std::forward_as_tuple(&parameterNode1),
-                    std::forward_as_tuple(false, true));
+                    std::forward_as_tuple(TIME_STRUCTURE_TYPE::VARYING_IN_SCENARIO_ONLY));
 
     // Variable time varying but constant across scenarios
     VariableNode variableNode1("v1");
     context.emplace(std::piecewise_construct,
                     std::forward_as_tuple(&variableNode1),
-                    std::forward_as_tuple(true, false));
+                    std::forward_as_tuple(TIME_STRUCTURE_TYPE::VARYING_IN_TIME_ONLY));
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp]:
```cpp
context[&variableNode1] = TIME_STRUCTURE_TYPE::VARYING_IN_TIME_ONLY;
```

**Review by flomnes — COMMENTED** [review]
1. Rename `TIME_STRUCTURE_TYPE` -> `TimeIndex`
2. Add a few more UT on `TimeIndex`, especially `operator|`
3. Apply clang-format

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h:1]: It seems that the `enum class` and `operator|` are defined twice (TimeIndex.h and TimeStructre.h), keep only one

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndexVisitor.h [hunk 4]
```diff
@@ -21,34 +21,33 @@
...
 namespace Antares::Solver::Visitors
 {
 
-class TimeIndexVisitor: public Nodes::NodeVisitor<TIME_STRUCTURE_TYPE>
+class TimeIndexVisitor: public Nodes::NodeVisitor<TimeIndex>
 {
 public:
-    using Base = Nodes::NodeVisitor<TIME_STRUCTURE_TYPE>;
-    TimeIndexVisitor(std::unordered_map<const Nodes::Node*, TIME_STRUCTURE_TYPE> context);
+    using Base = Nodes::NodeVisitor<TimeIndex>;
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndexVisitor.h]: Unused

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+
+namespace Antares::Solver::Visitors
+{
+enum class TimeIndex : int
+{
+    CONSTANT = 0,
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h]: `CONSTANT_IN_TIME_AND_SCENARIO`

Sorry for the multiple iterations on this enum, this is the last one ;)

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h (hunk 1 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h]: done

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h (hunk 2 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/TimeStructure.h]: done

File: src/tests/src/solver/expressions/test_expressions.cpp (hunk 3 above)
Comment (a-zakir) [reply src/tests/src/solver/expressions/test_expressions.cpp]: done

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h:1]: thank you

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h (hunk 5 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h]: updated

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+{
+enum class TimeIndex : unsigned int
+{
+    CONSTANT_IN_TIME_AND_SCENARIO = 0,
+    VARYING_IN_TIME_ONLY = 1,
+    VARYING_IN_SCENARIO_ONLY = 2,
+    VARYING_IN_TIME_AND_SCENARIO = 3
+};
+
+inline TimeIndex operator|(const TimeIndex& left, const TimeIndex& right)
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h]:
```suggestion
constexpr TimeIndex operator|(const TimeIndex& left, const TimeIndex& right)
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndexVisitor.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+#include "antares/solver/expressions/visitors/TimeIndex.h"
+
+namespace Antares::Solver::Visitors
+{
+
+class TimeIndexVisitor: public Nodes::NodeVisitor<TimeIndex>
+{
+public:
+    TimeIndexVisitor(std::unordered_map<const Nodes::Node*, TimeIndex> context);
+    TimeIndexVisitor() = default;
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndexVisitor.h]:
```suggestion
```

This visitor kind of needs a context, it's useless without one

File: src/solver/expressions/nodes/ParameterNode.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    Leaf(value),
+    TimeIndex()
+{
+}
+
+ParameterNode::ParameterNode(const std::string& value, bool time_varying, bool scenario_varying):
+    Leaf(value),
+    TimeIndex(time_varying, scenario_varying)
+{
+}
```
Comment (flomnes) [inline src/solver/expressions/nodes/ParameterNode.cpp]: You should not put anything related to `TimeIndex` into nodes

Comment (a-zakir) [reply src/solver/expressions/nodes/ParameterNode.cpp]: i've deleted this file, may be the fact that i changed the base branch has to do with it

File: src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+{
+    CONSTANT_IN_TIME_AND_SCENARIO = 0,
+    VARYING_IN_TIME_ONLY = 1,
+    VARYING_IN_SCENARIO_ONLY = 2,
+    VARYING_IN_TIME_AND_SCENARIO = 3
+};
+
+constexpr TimeIndex operator|(const TimeIndex& left, const TimeIndex& right)
+{
+    return static_cast<TimeIndex>(static_cast<int>(left) | static_cast<int>(right));
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/TimeIndex.h]:
```suggestion
    return static_cast<TimeIndex>(static_cast<unsigned int>(left) | static_cast<unsigned int>(right));
```

**Review by flomnes — APPROVED** [review]

**Pull request merged**